vaadin-dashboard-flow-parent/vaadin-dashboard-flow-integration-tests/src/test/java/com/vaadin/flow/component/dashboard/tests/DashboardDragDropIT.java

```diff
@@ -0,0 +1,94 @@
+public class DashboardDragDropIT extends AbstractComponentIT {
+
+    private DashboardElement dashboardElement;
+
+    @Before
+    public void init() {
+        open();
+        dashboardElement = $(DashboardElement.class).waitForFirst();
+    }
+
+    @Test
+    public void reorderWidgetOnClientSide_itemsAreReorderedCorrectly() {
+        DashboardWidgetElement widgetToReorder = dashboardElement.getWidgets()
+                .get(0);
+        reorderRootLevelItem(1, 0);
+        Assert.assertEquals(widgetToReorder.getTitle(),
+                dashboardElement.getWidgets().get(1).getTitle());
+    }
+
+    @Test
+    public void reorderSectionOnClientSide_itemsAreReorderedCorrectly() {
+        DashboardSectionElement sectionToReorder = dashboardElement
+                .getSections().get(0);
+        reorderRootLevelItem(2, 3);
+        Assert.assertEquals(sectionToReorder.getTitle(),
+                dashboardElement.getSections().get(1).getTitle());
+    }
+
+    @Test
+    public void reorderWidgetInSectionOnClientSide_itemsAreReorderedCorrectly() {
+        DashboardSectionElement firstSection = dashboardElement.getSections()
+                .get(0);
+        DashboardWidgetElement widgetToReorder = firstSection.getWidgets()
+                .get(1);
+        reorderWidgetInSection(2, 0, 1);
+        firstSection = dashboardElement.getSections().get(0);
+        Assert.assertEquals(widgetToReorder.getTitle(),
+                firstSection.getWidgets().get(0).getTitle());
+    }
+
+    @Test
+    public void detachReattach_reorderWidgetOnClientSide_itemsAreReorderedCorrectly() {
+        clickElementWithJs("toggle-attached");
+        clickElementWithJs("toggle-attached");
+        dashboardElement = $(DashboardElement.class).waitForFirst();
+        reorderWidgetOnClientSide_itemsAreReorderedCorrectly();
+    }
+
+    private void reorderWidgetInSection(int sectionIndex, int initialIndex,
+            int targetIndex) {
+        executeScript(
+                """
+                        const sectionIndex = %d;
+                        const reorderedItem = arguments[0].items[sectionIndex].items.splice(%d, 1)[0];
+                        arguments[0].items[sectionIndex].items.splice(%d, 0, reorderedItem);
+                        arguments[0].dispatchEvent(new CustomEvent('dashboard-item-reorder-end'));"""
+                        .formatted(sectionIndex, initialIndex, targetIndex),
+                dashboardElement);
+    }
+
+    private void reorderRootLevelItem(int initialIndex, int targetIndex) {
+        executeScript(
+                """
+                        const reorderedItem = arguments[0].items.splice(%d, 1)[0];
+                        arguments[0].items.splice(%d, 0, reorderedItem);
+                        arguments[0].dispatchEvent(new CustomEvent('dashboard-item-reorder-end'));"""
+                        .formatted(initialIndex, targetIndex),
+                dashboardElement);
```

Comment:
Could consider using Selenium for DnD. I'm fine with the JS approach also but it feels somewhat hacky.
```java
var initialWidget = dashboardElement.getWidgets().get(initialIndex);
var dragHandle = initialWidget.$("*").withClassName("drag-handle").first();

var targetWidget = dashboardElement.getWidgets().get(targetIndex);

var yOffset = initialWidget.getLocation().getY() < targetWidget.getLocation().getY() ? 10 : -10;
var xOffset = initialWidget.getLocation().getX() < targetWidget.getLocation().getX() ? 10 : -10;

new Actions(driver).clickAndHold(dragHandle)
        .moveToElement(targetWidget, xOffset, yOffset)
        .release(targetWidget)
        .build()
        .perform();
```

Reply:
Done. Also added tests for dragging items when editable is toggled.